Add main run loop, blackboard, and skeleton of modules

ah-OOG-ah opened the pull request:
What it says on the tin. It also strips out some manual logging in favor of the logging module.

ah-OOG-ah committed:
Add main run loop and partial modules

ah-OOG-ah committed:
Fix physics

jebent521 commented on server/index.js:
```diff
@@ -20,5 +27,92 @@ wss.on('connection', function connection(ws) {
+  const THRUST = 45_040;
+  const MASS_FLOW = THRUST / G_0 / I_SP;
+
+  if (!blackboard.get("position")) { blackboard.set("position", 15_000 + LUNAR_RADIUS); }
```
We should add an "altitude" field that's relative to the surface of the moon

jebent521 commented on server/index.js:
Currently it's possible to fall through the surface of the moon. Can you add something that stops the simulation when position <= lunar radius?

jebent521 commented on server/index.js:
```diff
@@ -20,5 +27,92 @@ wss.on('connection', function connection(ws) {
+    controlsMod(blackboard, isBurning);
+    physicsMod(blackboard);
+    loggingMod(blackboard);
+
```
This could either be a new module or part of the loggingMod, but we should be sending altitude, velocity, and fuel remaining back to the client at each tick (instead of sending thruster on/off)

ah-OOG-ah committed:
Add crashing, client updating, landing

jebent521 commented on server/index.js:
```diff
@@ -20,5 +27,125 @@ wss.on('connection', function connection(ws) {
+/**
+ * TALK TO THE CLIENT
+ */
+function communicationMod(blackboard, ws) {
```
This should be sent as a single map instead of multiple little sends, since the client will need to parse the map as a whole

ah-OOG-ah replied:
Why? For example, the blackboard contains `position`, which the server needs to calculate gravity but the client doesn't.

jebent521 replied:
Sorry I don't think I was clear. I agree the server should send _everything_ to the client. But what it _does_ send, it should send in the form of a map, so that way the client can parse it altogether and update the UI

jebent521 commented on server/index.js:
```diff
@@ -1,16 +1,23 @@
     if (message == 'thruster on') {
-      console.log('Thruster is on 🔥');
+      isBurning = true;
       ws.send('Thruster is on 🔥');
```
Now that you've got the communicationsMod in, you can get rid of these `send` statements

ah-OOG-ah replied:
The client still needs them, as the state of the thrusters is not necessarily linked to what the client is doing. The thruster state only changes on server ticks, and if the player is fast enough they could desync with the server unless we explicitly tell them the state.

ah-OOG-ah replied:
whoops, misread. Yeah, I'll remove these (and update the client accordingly)

jebent521 commented on server/index.js:
```diff
@@ -1,16 +1,23 @@
+      isBurning = false;
       ws.send('Thruster is off 💨');
     } else {
       console.log(`Unknown message: ${message}`);
```
I'd probably keep this one though - it'd be useful for debugging

ah-OOG-ah replied:
Logger already dumps the whole blackboard, which includes this

jebent521 replied:
The message isn't stored in the blackboard though. I'm imagining a debugging case where the client tries sending a message unknown to the server. We'd want to log it and send it back to the client for ease of debugging

ah-OOG-ah replied:
The "Unknown message" logger will stay, I'm just removing the thruster ones

jebent521 commented on server/index.js:
```diff
@@ -20,5 +27,125 @@ wss.on('connection', function connection(ws) {
+  const WARN_VEL = -3;
+  const KILL_VEL = -5;
+
+  if (!blackboard.has("position")) { blackboard.set("position", 15_000 + LUNAR_RADIUS); }
```
I'm allergic to embedded literals. For consistency and reliability, could you turn these into const `positionKey` etc. variables? I'm worried I'll accidentally type "poistion" or something goofy like that

jebent521 replied:
Not a big deal, we can address at a later date. I'd like this branch to be merged in ASAP

ah-OOG-ah committed:
Update client for thruster data

ah-OOG-ah committed:
Remove unused sends

ah-OOG-ah committed:
Use proper names

Also add time acceleration, for debugging
Also fix fuel consumption and crashing

ah-OOG-ah committed:
Better time acceleration

ah-OOG-ah committed:
No flying when you're dead

ah-OOG-ah committed:
Default time is realtime
